Hubspot doc refactor (#907)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/hubspot/hubspot-device-mode.mdx

</div>

### Dates

For sending properties of `date` type, you can send them as the epoch time or as a `date` object. RudderStack converts them to the required HubSpot format (midnight UTC).

## Track

A <Link to="/event-spec/standard-events/track/">`track`</Link> call lets you record various user actions and any properties associated with them.

To associate a `track` call with a user, you need to specify the user's `email` under `context.traits` . Additionally, RudderStack associates the `track` events with the same user after an `identify` request is successfully made.

A sample `track` event is as shown:

```javascript
rudderanalytics.track(
  "Order Completed", {
    value: 30,
  }, {
    context: {
      traits: {
        firstname: "Alex",
        city: "New Orleans",
        country: "USA",
        phone: "+1-202-555-0146",
        email: "alex@example.com",
      },
    },
  }
)
```

The `track` call also supports <Link to="https://legacydocs.hubspot.com/docs/methods/enterprise_events/http_api#:~:text=%26favorite_color%3Dorange-,External%20id,-%26id%3D%7Bvalue%7D">`externalId`</Link> as a parameter which is mapped from `externalID` `hubspotId`.

> 

Reviewer: example for page call
```
  rudderanalytics.page({
    path: "path",
  });
```

Author: Thanks, Akash!

---
title: "HubSpot Device Mode (Legacy API)"
description: Detailed technical documentation on sending events to HubSpot via device mode using the legacy API.
---

RudderStack supports the following API calls while sending data to Hubspot via web <Link to="/destinations/rudderstack-connection-modes/#device-mode">device mode</Link> using the [legacy API](https://legacydocs.hubspot.com/docs/overview?_ga=2.34803302.670362313.1663315856-97001172.1658910392).

<div class="infoBlock">
Find the open source code for HubSpot's web device mode in the <a href="https://github.com/rudderlabs/rudder-sdk-js/tree/production/integrations/HubSpot">GitHub repository</a>.
</div>

## Identify

RudderStack sends the <Link to="/event-spec/standard-events/identify">`identify`</Link> call to create or update a contact in HubSpot. The contact's email ID **must be present** in the `traits` object of the `identify` call.

> You must call either <code class="inline-code">page</code> or <code class="inline-code">track</code> after the <code class="inline-code">identify</code> call to create a contact. The previously made <code class="inline-code">page</code> and <code class="inline-code">track</code> calls will automatically be associated with that contact once they are identified.

Reviewer 2: ```suggestion
You must call either <code class="inline-code">page</code> or <code class="inline-code">track</code> after the <code class="inline-code">identify</code> call to create a contact. RudderStack automatically associates the previously made <code class="inline-code">page</code> and <code class="inline-code">track</code> calls with that contact once they are identified.
```

A sample `identify` call is as shown:

```javascript
rudderanalytics.identify({
  firstName: "Alex",
  city: "New Orleans",
  country: "USA",
  phone: "+1-202-555-0146",
  email: "alex@example.com",
  custom_flavor: "chocolate",
  custom_date: 1574769933368,
  custom_date1: new Date("2019-10-14T11:15:53.296Z"),
})

> ```

Reviewer 2: Do we need the `custom_` attributes in this sample call? It will be helpful to keep only the required/necessary details and keep the example simple, to avoid any confusion. 

Also, we should add a user ID to maintain consistency with other the sample `identify` calls. WDYT? @akashrpo @ishachopra0401

Reviewer: This sample call with custom parameters demonstrates the Custom properties which HS takes and that is explained below. userId is not a required field here.

Author: Thanks, Akash. It helps!

RudderStack then maps the field values present in the `traits` to the corresponding HubSpot custom property.

<div class="warningBlock">
HubSpot discards any property that does not exist and returns a <strong>400 Bad Request</strong> error.
</div>

### Dates

For sending properties of `date` type, you can send them as the epoch time or as a `date` object. RudderStack converts them to the required HubSpot format (midnight UTC).

## Track

A <Link to="/event-spec/standard-events/track/">`track`</Link> call lets you record various user actions and any properties associated with them.

To associate a `track` call with a user, you need to specify the user's `email` under `context.traits` . Additionally, RudderStack associates the `track` events with the same user after an `identify` request is successfully made.

A sample `track` event is as shown:

```javascript
rudderanalytics.track(
  "Order Completed", {
    value: 30,
  }, {
    context: {
      traits: {
        firstname: "Alex",
        city: "New Orleans",
        country: "USA",
        phone: "+1-202-555-0146",
        email: "alex@example.com",
      },
    },
  }
)
```

> The `track` call also supports <Link to="https://legacydocs.hubspot.com/docs/methods/enterprise_events/http_api#:~:text=%26favorite_color%3Dorange-,External%20id,-%26id%3D%7Bvalue%7D">`externalId`</Link> as a parameter which is mapped from `externalID` `hubspotId`.

Reviewer 2: Would it be helpful to demonstrate a sample `track` call containing the `externalId` parameter? @akashrpo

Reviewer: This sentence doesn't apply for HS device mode we should remove it.

### Revenue events

For the revenue events, a `value` or `revenue` key should be included in the event properties to be recorded in HubSpot.

## Page

You can make a <Link to="/event-spec/standard-events/page/">`page`</Link> call to automatically record page views.

A sample `page` event is as shown:

```javascript
rudderanalytics.page({
  path: "path",
});
```

## FAQ

### Can I use HubSpot website analytics?

> Yes - RudderStack supports sending `page` calls in device mode which can be used for HubSpot website analytics.

Reviewer 2: ```suggestion
Yes, you can. RudderStack supports sending `page` calls in device mode which can be used for HubSpot website analytics.
```
